File: scala/mergesort/Sort.scala
package mergesort{
  import scala.collection.mutable.Queue;
  import java.util.Comparator;
  import java.lang.Integer;
  import java.io.FileWriter;

  object Sort{
      val ASCENDING = 0;
      val DESCENDING = 1;

      def _mergeSort[T] (left : Integer, right : Integer, array : Array[T], 
        comparator : Comparator[T], ordering : Integer): Array[T] = {
          val middle = (left + right) / 2;
          val middleRight = middle + 1;

          if(left < right){
              _mergeSort(left, middle, array, comparator, ordering);
              _mergeSort(middleRight, right, array, comparator, ordering);
              _merge(left, middle, middle + 1, right, array, comparator, ordering);
              return array;
          }else{
              return array;
          }
      }

      def _merge[T] (lefti : Integer, 
        middlei : Integer, rightMiddlei : Integer, righti : Integer, 
        array : Array[T], comparator : Comparator[T], ordering : Integer): Array[T] = {
          var left = lefti;
          var middle = middlei;
          var rightMiddle = rightMiddlei;
          var right = righti;
          var initial = left;
          var arraySize = right - left + 1;
          val tempArray:Queue[T] = Queue();
          for(i <- 0 until arraySize){
              if(left > middle && rightMiddle <= right){
                  tempArray += array(rightMiddle);
                  rightMiddle += 1;
              }else if(rightMiddle > right && left <= middle){
                  tempArray += array(left);
                  left += 1;
              }else{
                  if(ordering == DESCENDING){
                      if(comparator.compare(array(left), array(rightMiddle)) < 0){
                          tempArray += array(left);
                          left += 1;
                      }else{
                          tempArray += array(rightMiddle);
                          rightMiddle += 1;
                      }
                  }else{
                      if(comparator.compare(array(left), array(rightMiddle)) < 0){
                          tempArray += array(rightMiddle);
                          rightMiddle += 1;
                      }else{
                          tempArray += array(left);
                          left += 1;
                      }
                  }
              }
          }
          for (i <- 0 until arraySize){
            val something = tempArray.dequeue();
              array(initial + i) = something;
          }
          return array;
      }
      
      def SortThis[T](comparator : Comparator[T], unSortedArray : Array[T], 
        ordering : Integer )(implicit m: reflect.Manifest[T]): Array[T] = {

          var orderingString = "Descending ";
          if(ordering == ASCENDING){
            orderingString = "Ascending ";
          }
          print(String.format("%-20s | %8s%n", orderingString + m.toString.split('.').last, 
            (time
              (_mergeSort(0, unSortedArray.length -1, unSortedArray, comparator, ordering)))
                .toString));
          return unSortedArray;
      }

    def time(f: => Unit): Long ={
      val s = System.nanoTime;
      f
      val duration = System.nanoTime - s;
      val fw = new FileWriter("scala.csv", true) ; 
      fw.write(duration.toString + ","); 
      fw.close()
      return duration;
    }


  }
}

File: java/mergesort/Sort.java
package mergesort;
import java.io.FileWriter;
import java.util.Comparator;

public class Sort{
    public final static int ASCENDING = 0;
    public final static int DESCENDING = 1;

    private <T> T[] _mergeSort(int left, int right, T[] array, Comparator comparator, int ordering){
        int middle = (left + right) / 2;
        int middleRight = middle + 1;

        if(left < right){
            _mergeSort(left, middle, array, comparator, ordering);
            _mergeSort(middleRight, right, array, comparator, ordering);
            _merge(left, middle, middle + 1, right, array, comparator, ordering);
            return array;
        }else{
            return array;
        }
    }

    private <T> void _merge(int left, int middle, int rightMiddle, int right, 
            T[] array, Comparator comparator, int ordering){
        int initial = left;
        int arraySize = right - left + 1;
        T[] tempArray = (T[])new Object[arraySize];
        for(int i = 0; i < tempArray.length; i++){
            if(left > middle && rightMiddle <= right){
                tempArray[i] = array[rightMiddle];
                rightMiddle++;
            }else if(rightMiddle > right && left <= middle){
                tempArray[i] = array[left];
                left++;
            }else{
                if(ordering == DESCENDING){
                    if(comparator.compare(array[left], array[rightMiddle]) < 0){
                        tempArray[i] = array[left];
                        left++;
                    }else{
                        tempArray[i] = array[rightMiddle];
                        rightMiddle++;
                    }
                }else{
                    if(comparator.compare(array[left], array[rightMiddle]) < 0){
                        tempArray[i] = array[rightMiddle];
                        rightMiddle++;
                    }else{
                        tempArray[i] = array[left];
                        left++;
                    }
                }
            }
        }
        for (int i = 0; i < tempArray.length; i++){
            array[initial + i] = tempArray[i];
        }
    }
    
    public <T> T[] SortThis(Comparator comparator, T[] unSortedArray, int ordering ){
      String orderingString = "Descending ";
      if(ordering == ASCENDING){
        orderingString = "Ascending ";
      }
      Long s = System.nanoTime();

        _mergeSort(0, unSortedArray.length -1, unSortedArray, comparator, ordering);
      Long duration = System.nanoTime() - s;
      try{
      FileWriter fw = new FileWriter("java.csv", true) ; 
      fw.write(duration.toString() + ","); 
      fw.close();
      }catch(Exception e){
      }
        
        return unSortedArray;
    }

}
